EasyLM/models/llama/llama_train_ppo.py

```diff
@@ -0,0 +1,421 @@
+    # input_ids, attention_mask, continuation_mask = batch['prompt_input_ids'], batch['prompt_attn_mask'], batch['prompt_attn_mask']
+    generation_config = GenerationConfig(
+        max_new_tokens=32,
+        do_sample=True,
+        temperature=1.0,
+    )
+    pad_token_id = 0
+    outputs = policy_model.generate(
+        input_ids=batch['prompt_input_ids'],
+        generation_config=generation_config,
+        params=policy_train_state.params,
+        pad_token_id=pad_token_id,
+    )
+    # print(outputs.sequences)
+    input_ids = outputs.sequences
+    attention_mask = jnp.where(input_ids == pad_token_id, 0, 1)
+    continuation_mask = jnp.concatenate([
+        jnp.zeros_like(batch['prompt_attn_mask']),
+        jnp.ones_like(attention_mask[:, batch['prompt_attn_mask'].shape[1]:]),
+    ], axis=1)
+
+    # run forward pass on policy, output `continuations_logits` (B, L, V) and `continuations_logprobs` (B, L)
+    continuations_logits = policy_model(input_ids, attention_mask, params=policy_train_state.params).logits
+    continuations_logps = convert_logits_to_logps(continuations_logits, input_ids, continuation_mask) # (B, L)
+    continuations_logps = jax.lax.stop_gradient(continuations_logps)
+
+    # run forward pass on reference, output `continuations_ref_logits` (B, L, V) and `continuations_ref_logprobs` (B, L)
+    continuations_ref_logits = reference_model(input_ids, attention_mask, params=reference_train_state.params).logits
+    continuations_ref_logps = convert_logits_to_logps(continuations_ref_logits, input_ids, continuation_mask) # (B, L)
+    continuations_ref_logps = jax.lax.stop_gradient(continuations_ref_logps)
+
+    # run forward pass on value, output `continuations_value` (B, L)
+    continuations_value = value_model(input_ids, attention_mask, params=value_train_state.params).logits[:, :, 0]
+    continuations_value = jax.lax.stop_gradient(continuations_value)
+
+    # run reward model, output `rewards_raw` (B), `rewards_normalized` (B), `rewards_kl` (B, L), `rewards_kl_penalty` (B, L), `rewards_penalized` (B, L)
+    rewards = reward_model(input_ids, attention_mask, params=reward_train_state.params).logits[:, :, 0]
+    last_token_index = attention_mask.sum(axis=-1) - 1 # (B)
+    rewards_raw = jnp.take_along_axis(rewards, last_token_index[:, None], axis=1).squeeze(-1) # (B)
+    rewards_normalized = rewards_raw * FLAGS.reward_gain + FLAGS.reward_bias # (B)
+    rewards_flattened = jnp.zeros_like(rewards)
+    rewards_flattened = rewards_flattened.at[:, last_token_index].set(rewards_normalized) # (B, L)
+    rewards_kl = continuations_logps - continuations_ref_logps # (B, L)
+    rewards_kl_penalty = FLAGS.kl_coef * rewards_kl # (B, L)
+    rewards_penalized = rewards_flattened - rewards_kl_penalty # (B, L)
+    rewards_penalized = jax.lax.stop_gradient(rewards_penalized)
+
+    for ministep in range(FLAGS.ministeps):
```

Comment:
How do people usually schedule LRs for PPO with the ministep setup? In flax/optax, the step count in the train state (updated every time there is an `apply_gradients` call) is usually used to work out where the LR should be, so we might have to be careful about this. The way gradient accumulation also works is that it just updates params every time `step % accumulation_steps == 0`, which can mean the train_state steps doesn't reflect the number of gradient steps (and also means if ministeps < accumulation steps, then the model won't get updated).

Reply:
I think for PPO we should get rid of the `accumulation_steps` and update every step. As for LR, AFAIK people tend to use a constant LR, so we don't need to worry too much about tracking exact steps.

Reply:
Easy, thats good! Then we need to set some stuff in the optimizer to make it a constant LR - init_lr and end_lr should be the same